data/blog/boosting-mobile-session-replay-performance-with-view-renderer-v2.md

```diff
@@ -0,0 +1,312 @@
+
+If the workload is below that, e.g. `4 ms`, all is good and we wait for the next update. But if the workload takes longer than that, e.g. `25ms`, we have a problem — the main thread is blocked and the app hangs.
+
+As animations are not coupled to frames but to time, the app cannot simply continue with rendering the next one afterwards. For example, if you expect a loading indicator to rotate exactly once per second, longer frame render times would slow down the animation.
+
+To correct the timing the system **skips** the frames that should already have been rendered at this point in time (the _dropped frames)_. This correction behavior is commonly known as _frame drops_ and has the undesired side effect of intermittent frame rate changes that can be noticed as visual stutters (i.e. not scrolling smoothly).
+
+<div align="center">
+  <img src="/images/boosting-mobile-session-replay-performance-with-view-renderer-v2/frame-drops.jpg" alt="Dropped Frames"/>
+  _When workload takes longer than available time, we need to drop frames_
+</div>
+
+Frame drops were also the unintended side effect for the previous implementation of Session Replay, as the once-per-second screenshot simply took too long.
+
+# Taking Screenshots With PII In Mind
+
+Now that we know what the issue is, we need to take a closer look at the work done by Session Replay. The capture process consists of three steps that must be executed for each screenshot:
+
+1. **Redact**: Traverse the view hierarchy to identify UI elements that contain sensitive information and require masking.
+2. **Render**: Convert the current view hierarchy into a rasterized bitmap representation ("take a screenshot") using a _view renderer_.
+3. **Mask**: Apply privacy masking by combining the rendered image with the redaction geometry.
+
+As mentioned before, accessing the view hierarchy must be done on the main thread, which means that both steps _Redact_ and _Render_ must also execute on the main thread. Only the final step _Mask_ can be performed on a background queue, since it operates solely on the image data produced by the previous steps.
+
+This allows us to focus on the performance of the _Redact_ and _Render_ steps.
+
+<div align="center">
+  <img src="/images/boosting-mobile-session-replay-performance-with-view-renderer-v2/uikit-meme.png" alt="UIKit Meme"/>
+  _Source: [@ios_memes on X](https://x.com/ios_memes/status/1483418141871030276)_
+</div>
+
+# Measuring Baseline Performance
+
+After adding some code to calculate the execution time and adding a basic sampling mechanism to reduce variation over time, we quickly gathered our first insights using a (rather old) iPhone 8:
+
+**The step _Render_ takes `~155ms` per frame, causing ~9-10 of 60 frames being dropped every second of using Session Replay** 💀
+
+| 120 samples | Redact        | Render          | Total           |
```

Comment:
The totals don't quite add up, but I'm assuming this is because I can only see/add up 4 decimal places?